### ISSUE TITLE
Class hidden properties still serialized to JSON

### ISSUE THREAD

**Issue by Brad-Christie** [PowerShell/PowerShell#9847]
Any reason why `hidden` has no bearing on `ConvertTo-Json`? Seems that `Get-Member` honors visibility, but JSON serialization does not.

# Steps to reproduce

```powershell
Class Person {
  hidden [string] $FirstName
  hidden [string] $LastName

  [string] $FullName

  Person([string]$First, [string]$Last) {
    $this.FirstName = $First
    $this.LastName = $Last

    $this.FullName = $First, $Last -join " "
  }

  [string] ToString() {
    Return $this.FullName
  }
}

$jDoe = [Person]::new("John", "Doe")

Write-Output "`nProperties:"
$jDoe | Get-Member -MemberType "Property" | Format-Table "Name", "MemberType"

Write-Output "`nJSON:"
$jDoe | ConvertTo-Json
```

# Expected behavior

```none
Properties:

Name     MemberType
----     ----------
FullName   Property


JSON:
{
  "FullName": "John Doe"
}
```

# Actual behavior

```none
Properties:

Name     MemberType
----     ----------
FullName   Property


JSON:
{
  "FirstName": "John",
  "LastName": "Doe",
  "FullName": "John Doe"
}
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.2.1
PSEdition                      Core
GitCommitId                    6.2.1
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

FWIW, Appears `ConvertTo-Xml -As "String"` _does_ honor visibility:

```none
<?xml version="1.0" encoding="utf-8"?>
<Objects>
  <Object Type="Person">
    <Property Name="FullName" Type="System.String">John Doe</Property>
  </Object>
</Objects>
```

Also, did see #6652 and there's some mention re: hidden with regards to format-* but didn't see anythign related to JSON serialization.

**Comment by Brad-Christie** [comment]
Looked uickly, appears XML uses a `CustomSerialization` serializer, which appears to [iterate the PSObject.properties](https://github.com/PowerShell/PowerShell/blob/aba3bd5df1ea9d97790e7b232e50995e9f3d1e37/src/Microsoft.PowerShell.Commands.Utility/commands/utility/CustomSerialization.cs#L608) itself, which I assume implements some kind of filtering mechanism (through the `PSMemberInfoIntegratingCollection<>`). [`ConvertTo-Json`](https://github.com/PowerShell/PowerShell/blob/50efc4192d99bbbe20c34e03ab04aa2f7d7d82b1/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs), however, appears to [hand it off](https://github.com/PowerShell/PowerShell/blob/50efc4192d99bbbe20c34e03ab04aa2f7d7d82b1/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs#L128) to newtonsof without it benefiting from context (which, I suspect, just uses reflection).

Looks like you'd have to create a ContractResolver and pass it off that context for it to work the same, at least from what I gathered from doing a preliminary. 🤷‍♂️

**Comment by thezim** [comment]
Someone correct me if I’m wrong but isn’t serialization/deserialization a misnomer for the Convert* cmdlets? You are never going to get a functional class back out of a “serialized” class unless it a POCO to start with. It’s something different by then in the example here.

**Comment by Brad-Christie** [comment]
@thezim I can appreciate that, but not looking to convert back. The example was handy given I ran into the issue, but the original intent is to use a model to build a final object. That object is then consumed by an external system, and I'd rather not expose the intermediary properties.

That said, It's still inconsistent behavior. Get-Member and Csv, Html & Xml work, so (IMHO) at least ConvertTo-Json should work too.

**Comment by mklement0** [comment]
Just to affirm that the current behavior is indeed unexpected:

`Select-Object` (with wildcards) and `ConvertTo-Csv` / `Export-Csv` - sensibly - do _not_ include hidden properties - `ConvertTo-Json` should act the same.

(`Export-CliXml` does, but there it makes sense (as-faithful-as-possible serialization); the rarely used `ConvertTo-Xml` does _not_).

**Comment by iSazonov** [comment]
May be fixed in #11198 as a side effect. :-)

**Comment by dbramblett** [comment]
I just ran into this as a problem as well. I create objects which turned out to have some hidden properties, which are in turn converted to json and exported.  Because ConvertTo-Json does not behave the same way as Get-Member, ConvertTo-Csv, etc., it took me longer than it should have to track down why a property that didn't appear to be part of any of the code anywhere I'd written, was popping up in the json output and creating problems when I tried to import that content back into powershell somewhere down the road. As @nosalan remarked, this is a bug and should be re-opened.
